Remove the special handling of exit code 2

joeytwiddle opened the pull request:
This PR removes the special handling of exit code 2.

Not only `mocha` but also `stylelint` will sometimes exit with code 2, resulting in unexpected behaviour from nodemon. (It exits, and stops monitoring.)

The ` || true` workaround is unusual and easy to forget. I only ever remember to add it *after* I have been bitten!

If there is [no special reason](https://github.com/remy/nodemon/issues/627#issuecomment-139053515) to keep the special handling of code 2, then I think it would be best to remove it.

(In the case when a developer does provide a non-parseable command, which exits with code 2, an error is already reported to the console. So we won't be keeping developers in the dark by not quitting. After this PR, the developer will need to hit Ctrl-C to exit nodemon, in order to fix their command. For an example syntax error, try: `nodemon -x 'ls <'`)

Fixes #496 #627

This could perhaps be part of a major release, just in case some people are using it as a way to break out of nodemon intentionally: https://github.com/remy/nodemon/issues/751#issuecomment-267431143

remy commented:
I'm not convinced this is the right change. I think the messaging in the exit should be clearer, but that's all.

With your changes, here is the workflow:

```
❯ nodemon -x 'ls <'
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `ls <`
sh: -c: line 0: syntax error near unexpected token `newline'
sh: -c: line 0: `ls <'
[nodemon] app crashed - waiting for file changes before starting...
```

Since the actual exec command is wrong, the developer has to `ctrl+c` nodemon and correct the exec command.

Automatically restarting on file changes will never make that command work.

I think the right change is to update the error message that's shown, perhaps linking to the FAQ (which can be added to with links and even more context).

I can't see a reason for nodemon _not_ to exit on code 2 - unless I'm missing something?

joeytwiddle commented:
The argument not to exit on code 2 is that mocha will sometimes exit with code 1, 2, 3, 4, ... (and stylelint too), and usually in those cases we still want `nodemon` to keep watching and running.

Currently behaviour with mocha is somewhat inconsistent.  `nodemon` will happily watch and run passing and failing tests for a while, but then it will surprisingly stop running entirely (when exactly 2 mocha tests fail). It appears random, until you find the GitHub issue explaining that code 2 is special, and you are advised to put `|| true` at the end of your mocha command.

On the other hand, yes sometimes code 2 means you wrote a bad shell command. And in that case exiting nodemon is slightly preferable.

So I guess we have to decide which is more bothersome: nodemon stops running mocha tests unexpectedly, or having to hit <kbd>Ctrl-C</kbd> after typing a bad command.

Anyway I do agree, in either case, some messaging may help to guide the developer. I will look into that.

Other options:
1. We could try harder to detect whether the code 2 is from a bad command or from mocha/stylelint. For example:
1a. we could read stderr for things like `sh: -c: line 0: syntax error`, or
1b. we could measure if the command fails _immediately_ the _first time_ it is executed.
2. We could raise issues on mocha and stylelint packages, explaining that they are using the reserved code 2 incorrectly. (But I suspect they will not want to change the behaviour.)

I would like to look into 1b, and situational messaging, when I have time.

joeytwiddle commented:
OK, I have changed the behaviour:
- If the exit code is 2, and the time taken for the child to fail is < 500ms, then we assume the command had a syntax error.
- I have improved the messaging a bit (perhaps verbose, but hopefully clear).
- (I have not checked if this is the first run or not. It doesn't seem necessary.)

To test, try:
```sh
./bin/nodemon.js -x 'ls <'

# and

./bin/nodemon.js -x 'sleep 1 ; exit 2'
```

remy commented on lib/monitor/run.js:
```diff
@@ -158,9 +158,18 @@ function run(options) {
+      utils.log.error('or it is exiting with reserved code 2.');
+      utils.log.error('');
+      utils.log.error('To keep nodemon running even after a code 2,');
+      utils.log.error('add this to the end of your command: || true');
```
Can you add: `Read more here: https://git.io/fNOAG` after line 169.

remy commented:
Two comments I've left, and once these are in, I'm happy with the PR - thank you and great work 👍

joeytwiddle commented:
Remy, I only saw one comment from you! I have added the link to the FAQ.

But what was the second comment? I have taken a guess, and changed the `|| true` into `|| exit 1` like the FAQ suggests.

Thank you for `nodemon`, an invaluable productivity tool! :smile:

remy commented:
I don't understand what github is doing, but I've taken a screenshot of the comments - sorry, bit naff!

<img width="977" alt="screen shot 2018-07-17 at 17 00 04" src="https://user-images.githubusercontent.com/13700/42830332-02036386-89e3-11e8-810a-e2953d3ce5b8.png">

joeytwiddle commented:
My apologies. I probably made a rebase/amend and a force push yesterday, just after you commented, so the comment got lost.

I have added a comment with the fNOAR link now.

remy commented:
> If you want to make adjustments to this branch directly, I will not be offended. 🙇

I've never really known how do that cleanly. Once I clone the branch locally, (from past experience) I couldn't push back into the PR and had to make my _own_ branch and new PR - really odd. Anyway, this looks good - thanks!
